# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

<img src="pandas_1.png">

*source: https://pandas.pydata.org/pandas-docs/stable/overview.html*

In this section, you will study:
1. The pandas Series (similar to a numpy array)
    * Creating a pandas series
    * Indexing series
2. Dataframes 
    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes 

In [17]:
import pandas as pd

train_data=pd.read_csv('train.csv')
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
print(train_data[["Sex", "Survived"]].groupby(['Sex'], as_index=False).count())

      Sex  Survived
0  female       314
1    male       577


In [4]:
x=train_data[(train_data.Sex=='male'), (train_data.Survived==1)]
x

TypeError: ignored

In [ ]:
x=train_data[((train_data.Sex=='male')|(train_data.Sex=='female')) & (train_data.Survived==1)]
x.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
8            9         1       3  ...  11.1333   NaN         S
9           10         1       2  ...  30.0708   NaN         C

[5 rows x 12 columns]

In [ ]:
xx=train_data[(train_data.Survived==1)].groupby('Sex').count()
ax = xx.plot.bar(x=['Male','Female'], y='A', rot=0)

KeyError: ignored

In [ ]:
print(train_data[["Sex", "Survived"]].groupby(['Sex'], as_index=False).mean().sort_values(by='Survived', ascending=False))

      Sex  Survived
0  female  0.742038
1    male  0.188908


### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [4]:
# import pandas, pd is an alias
import pandas as pd

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

s1 = pd.Series([2, 4, 5, 6, 9], index=[1,2,3,4,5])
print(s1)
print(type(s1))



0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>
1    2
2    4
3    5
4    6
5    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [5]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [6]:
# Nan values to series
import numpy as np
s1 = pd.Series([2, 4, 5,'np.nan'])
print(s1)
print(type(s1))


0         2
1         4
2         5
3    np.nan
dtype: object
<class 'pandas.core.series.Series'>


In [23]:
# creating a series of type datetime
#pd.date_range(date,period,frequency (d,M,Y))
date_series1=pd.date_range('20191110',periods=10,freq='D')
print(date_series1)
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
print(date_series)
type(date_series)

DatetimeIndex(['2019-11-10', '2019-11-11', '2019-11-12', '2019-11-13',
               '2019-11-14', '2019-11-15', '2019-11-16', '2019-11-17',
               '2019-11-18', '2019-11-19'],
              dtype='datetime64[ns]', freq='D')
DatetimeIndex(['2017-11-09', '2017-11-10', '2017-11-11', '2017-11-12',
               '2017-11-13', '2017-11-14', '2017-11-15', '2017-11-16',
               '2017-11-17', '2017-11-18', '2017-11-19', '2017-11-20',
               '2017-11-21', '2017-11-22', '2017-11-23', '2017-11-24',
               '2017-11-25', '2017-11-26', '2017-11-27', '2017-11-28',
               '2017-11-29', '2017-11-30', '2017-12-01', '2017-12-02',
               '2017-12-03', '2017-12-04', '2017-12-05', '2017-12-06',
               '2017-12-07', '2017-12-08', '2017-12-09', '2017-12-10',
               '2017-12-11', '2017-12-12'],
              dtype='datetime64[ns]', freq='D')


pandas.core.indexes.datetimes.DatetimeIndex

In [25]:
import numpy as np
r=np.random.randn(10,4)  # 10 rows and 4 cols
print(r)

df=pd.DataFrame(r)
print(df)    # printing data frame

df1=pd.DataFrame(r, index=date_series1, columns=list('ABCD'))
print(df1)

[[-0.63417201  0.51182108  0.15286448  0.34494309]
 [-0.57221346 -1.60321024  1.30361699  1.02461066]
 [-0.03370618 -0.26397132  0.62550196 -0.21138746]
 [ 0.52617279 -0.44739719 -0.42673275  0.95832201]
 [-0.49450296 -0.41380077  0.52395136 -0.67926867]
 [ 0.94948201 -0.51714559  1.03247298 -0.95053966]
 [-0.09114451 -0.95997745 -0.72508036 -0.31137232]
 [-1.29343884 -1.01787511 -0.73447337 -0.30723754]
 [-0.56386761 -0.5954133  -1.63867035 -0.2120124 ]
 [-0.79608789  0.19409608 -0.22468148 -1.73942384]]
          0         1         2         3
0 -0.634172  0.511821  0.152864  0.344943
1 -0.572213 -1.603210  1.303617  1.024611
2 -0.033706 -0.263971  0.625502 -0.211387
3  0.526173 -0.447397 -0.426733  0.958322
4 -0.494503 -0.413801  0.523951 -0.679269
5  0.949482 -0.517146  1.032473 -0.950540
6 -0.091145 -0.959977 -0.725080 -0.311372
7 -1.293439 -1.017875 -0.734473 -0.307238
8 -0.563868 -0.595413 -1.638670 -0.212012
9 -0.796088  0.194096 -0.224681 -1.739424
                   A       

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [ ]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]

# accessing elements starting index = 2 till the end
s[2:]

2    5
3    6
4    9
dtype: int64

In [ ]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [26]:
# creating data frames 

a=[['Deepti',10],['Ankit',20],['Aarti',25]]
df=pd.DataFrame(a)
print(a)
type(a)

a1=[['Deepti',10],['Ankit',20],['Aarti',25]]
df=pd.DataFrame(a1,columns=['Name','marks'])
print(df)
type(df)

[['Deepti', 10], ['Ankit', 20], ['Aarti', 25]]
     Name  marks
0  Deepti     10
1   Ankit     20
2   Aarti     25


pandas.core.frame.DataFrame

In [27]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

name  age    occupation
0   Vinay   22      engineer
1  Kushal   25        doctor
2    Aman   24  data analyst
3    Saif   28       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [39]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("market_fact.csv")


Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [ ]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id  ...   Profit  Shipping_Cost  Product_Base_Margin
0  Ord_5446  Prod_16  SHP_7609  ...   -30.51           3.60                 0.56
1  Ord_5406  Prod_13  SHP_7549  ...     4.56           0.93                 0.54
2  Ord_5446   Prod_4  SHP_7610  ...  1148.90           2.50                 0.59
3  Ord_5456   Prod_6  SHP_7625  ...   729.34          14.30                 0.37
4  Ord_5485  Prod_17  SHP_7664  ...  1219.87          26.30                 0.38

[5 rows x 10 columns]

In [ ]:
market_df.tail()

Ord_id  Prod_id   Ship_id  ...  Profit  Shipping_Cost  Product_Base_Margin
8394  Ord_5353   Prod_4  SHP_7479  ...  374.63           7.69                 0.59
8395  Ord_5411   Prod_6  SHP_7555  ...  -74.03           6.92                 0.37
8396  Ord_5388   Prod_6  SHP_7524  ...  -70.85           5.35                 0.40
8397  Ord_5348  Prod_15  SHP_7469  ...  565.34          30.00                 0.62
8398  Ord_5459   Prod_6  SHP_7628  ...  131.39           4.86                 0.38

[5 rows x 10 columns]

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [ ]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.3+ KB


In [40]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [41]:
#Describe using include object
market_df.describe(include=['object'])

Ord_id Prod_id  Ship_id    Cust_id
count      8399    8399     8399       8399
unique     5506      17     7701       1832
top     Ord_542  Prod_6  SHP_564  Cust_1140
freq          6    1225        4         30

In [42]:
#Describe using include object 
market_df.describe(include='all')

Ord_id Prod_id  Ship_id    Cust_id         Sales     Discount  \
count      8399    8399     8399       8399   8399.000000  8399.000000   
unique     5506      17     7701       1832           NaN          NaN   
top     Ord_542  Prod_6  SHP_564  Cust_1140           NaN          NaN   
freq          6    1225        4         30           NaN          NaN   
mean        NaN     NaN      NaN        NaN   1775.878179     0.049671   
std         NaN     NaN      NaN        NaN   3585.050525     0.031823   
min         NaN     NaN      NaN        NaN      2.240000     0.000000   
25%         NaN     NaN      NaN        NaN    143.195000     0.020000   
50%         NaN     NaN      NaN        NaN    449.420000     0.050000   
75%         NaN     NaN      NaN        NaN   1709.320000     0.080000   
max         NaN     NaN      NaN        NaN  89061.050000     0.250000   

        Order_Quantity        Profit  Shipping_Cost  Product_Base_Margin  
count      8399.000000   8399.000000    8399.000000          8336.000000  
unique             NaN           NaN            NaN                  NaN  
top                NaN           NaN            NaN                  NaN  
freq               NaN           NaN            NaN                  NaN  
mean         25.571735    181.184424      12.838557             0.512513  
std          14.481071   1196.653371      17.264052             0.135589  
min           1.000000 -14140.700000       0.490000             0.350000  
25%          13.000000    -83.315000       3.300000             0.380000  
50%          26.000000     -1.500000       6.070000             0.520000  
75%          38.000000    162.750000      13.990000             0.590000  
max          50.000000  27220.690000     164.730000             0.850000

In [31]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [32]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [46]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.values
market_df.value_counts()  # to count frequency of all values in columns
market_df['Prod_id'].value_counts()

Prod_6     1225
Prod_3      915
Prod_4      883
Prod_5      788
Prod_8      758
Prod_13     633
Prod_1      546
Prod_2      434
Prod_15     386
Prod_11     361
Prod_17     337
Prod_12     288
Prod_9      246
Prod_10     189
Prod_7      179
Prod_16     144
Prod_14      87
Name: Prod_id, dtype: int64

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [ ]:
market_df.head()

Ord_id  Prod_id   Ship_id  ...   Profit  Shipping_Cost  Product_Base_Margin
0  Ord_5446  Prod_16  SHP_7609  ...   -30.51           3.60                 0.56
1  Ord_5406  Prod_13  SHP_7549  ...     4.56           0.93                 0.54
2  Ord_5446   Prod_4  SHP_7610  ...  1148.90           2.50                 0.59
3  Ord_5456   Prod_6  SHP_7625  ...   729.34          14.30                 0.37
4  Ord_5485  Prod_17  SHP_7664  ...  1219.87          26.30                 0.38

[5 rows x 10 columns]

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [34]:
# Setting index to Ord_id
market_df.set_index('Ord_id', inplace = True)
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [35]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = False)

Prod_id   Ship_id   Cust_id     Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_999   Prod_15  SHP_1383  Cust_361   5661.08      0.00              33   
Ord_998    Prod_8  SHP_1380  Cust_372    750.66      0.00              33   
Ord_998    Prod_5  SHP_1382  Cust_372   2149.37      0.03              42   
Ord_998    Prod_8  SHP_1381  Cust_372    254.32      0.01               8   
Ord_997   Prod_14  SHP_1379  Cust_365  28761.52      0.04               8   
...           ...       ...       ...       ...       ...             ...   
Ord_1001   Prod_5  SHP_1385  Cust_374   1981.26      0.07              49   
Ord_1000   Prod_6  SHP_1384  Cust_373    334.71      0.01              25   
Ord_100    Prod_8   SHP_138   Cust_58    121.12      0.10               3   
Ord_10     Prod_3    SHP_13   Cust_10     80.61      0.02              15   
Ord_1      Prod_1     SHP_1    Cust_1    261.54      0.04               6   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_999   1055.47          30.00                 0.62  
Ord_998    120.05           4.00                 0.60  
Ord_998    217.87          19.99                 0.55  
Ord_998   -117.39           6.50                 0.79  
Ord_997    285.11          24.49                 0.37  
...           ...            ...                  ...  
Ord_1001   100.80           8.66                 0.76  
Ord_1000    31.74           6.47                 0.38  
Ord_100   -118.82           1.99                 0.44  
Ord_10      -4.72           2.99                 0.37  
Ord_1     -213.25          35.00                 0.80  

[8399 rows x 9 columns]

In [ ]:
# Sorting by values

# Sorting in increasing order of Sales
market_df.sort_values(by='Sales').head()

Prod_id   Ship_id  ... Shipping_Cost  Product_Base_Margin
Ord_id                       ...                                   
Ord_704    Prod_7   SHP_964  ...          0.70                 0.37
Ord_149    Prod_3  SHP_7028  ...          1.49                 0.37
Ord_4270   Prod_7  SHP_5959  ...          0.70                 0.81
Ord_4755  Prod_13  SHP_6628  ...          0.70                 0.56
Ord_2252   Prod_3  SHP_3064  ...          1.49                 0.37

[5 rows x 9 columns]

In [38]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55

In [ ]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id', 'Sales'], ascending = False)

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                     
Ord_2197  Prod_9  SHP_2994   Cust_827  7522.80      0.04              48   
Ord_4356  Prod_9  SHP_6074  Cust_1481  6831.72      0.01              41   
Ord_262   Prod_9   SHP_358    Cust_66  6553.45      0.03              39   
Ord_4059  Prod_9  SHP_5660  Cust_1378  5587.20      0.05              36   
Ord_2973  Prod_9  SHP_6073  Cust_1480  5410.95      0.09              36   
Ord_950   Prod_9  SHP_1315   Cust_334  4906.85      0.09              32   
Ord_5112  Prod_9  SHP_7141  Cust_1729  4273.95      0.05              49   
Ord_612   Prod_9   SHP_836   Cust_687  3872.38      0.10              50   
Ord_3443  Prod_9  SHP_4773  Cust_1246  3849.17      0.06              46   
Ord_3650  Prod_9  SHP_6795  Cust_1683  3353.54      0.07              22   
Ord_4590  Prod_9  SHP_6385  Cust_1465  3212.97      0.03              37   
Ord_2774  Prod_9  SHP_3809  Cust_1015  3191.24      0.00              38   
Ord_793   Prod_9  SHP_1089   Cust_245  2833.19      0.06              34   
Ord_3071  Prod_9  SHP_4265  Cust_1141  2728.42      0.06              45   
Ord_1290  Prod_9  SHP_1780   Cust_482  2609.53      0.00              27   
Ord_555   Prod_9   SHP_754   Cust_202  2568.71      0.10              43   
Ord_2853  Prod_9  SHP_3927  Cust_1056  2366.18      0.10              16   
Ord_4370  Prod_9  SHP_6092  Cust_1492  2325.42      0.02              25   
Ord_3239  Prod_9  SHP_4491  Cust_1205  2300.45      0.02              36   
Ord_2968  Prod_9  SHP_4092  Cust_1066  2160.83      0.08              36   
Ord_4989  Prod_9  SHP_6965  Cust_1700  2048.78      0.05              24   
Ord_4663  Prod_9  SHP_6498  Cust_1612  2026.42      0.03              50   
Ord_1152  Prod_9  SHP_1586   Cust_443  1554.53      0.00              16   
Ord_5370  Prod_9  SHP_7500  Cust_1809  1527.42      0.04              25   
Ord_3917  Prod_9  SHP_5439  Cust_1337  1504.01      0.05              42   
Ord_4595  Prod_9  SHP_6390  Cust_1523  1500.84      0.10              46   
Ord_3389  Prod_9  SHP_4698  Cust_1236  1484.89      0.03              41   
Ord_604   Prod_9   SHP_826   Cust_213  1406.64      0.02              50   
Ord_2737  Prod_9  SHP_3755   Cust_994  1400.91      0.09              49   
Ord_2461  Prod_9  SHP_3373   Cust_941  1396.75      0.04              46   
...          ...       ...        ...      ...       ...             ...   
Ord_1666  Prod_1  SHP_2303   Cust_490   101.52      0.02               6   
Ord_3693  Prod_1  SHP_5118  Cust_1287    96.03      0.10               6   
Ord_2133  Prod_1  SHP_2910   Cust_808    96.01      0.01              11   
Ord_3579  Prod_1  SHP_4956  Cust_1272    94.99      0.06              13   
Ord_1654  Prod_1  SHP_2284   Cust_491    90.58      0.06               7   
Ord_5479  Prod_1  SHP_7655  Cust_1805    88.94      0.05               5   
Ord_3506  Prod_1  SHP_4857   Cust_395    79.93      0.00               2   
Ord_1858  Prod_1  SHP_2820   Cust_720    73.86      0.08               7   
Ord_75    Prod_1    SHP_99    Cust_50    73.07      0.08               4   
Ord_5107  Prod_1  SHP_7132  Cust_1708    72.72      0.06               4   
Ord_2725  Prod_1  SHP_3736   Cust_997    70.02      0.00               4   
Ord_1933  Prod_1  SHP_2649   Cust_707    69.57      0.05               8   
Ord_4613  Prod_1  SHP_6420  Cust_1569    68.03      0.02               4   
Ord_3379  Prod_1  SHP_4685  Cust_1233    66.92      0.01               5   
Ord_542   Prod_1   SHP_735   Cust_188    63.84      0.02               3   
Ord_210   Prod_1   SHP_288    Cust_94    63.52      0.08               4   
Ord_4618  Prod_1  SHP_6432  Cust_1575    58.90      0.03               5   
Ord_212   Prod_1   SHP_291    Cust_88    58.33      0.02               3   
Ord_2410  Prod_1  SHP_3307   Cust_963    57.48      0.00               3   
Ord_4213  Prod_1  SHP_5875  Cust_1428    53.30      0.00 